xcrypto/hash_test.go

```diff
@@ -454,6 +454,44 @@ func TestHashAllocations(t *testing.T) {
 	}
 }
 
+func TestHashStructAllocations(t *testing.T) {
+	if Asan() {
+		t.Skip("skipping allocations test with sanitizers")
+	}
+	msg := []byte("testing")
+
+	md5Hash := xcrypto.NewMD5()
+	sha1Hash := xcrypto.NewSHA1()
+	sha256Hash := xcrypto.NewSHA256()
+	sha512Hash := xcrypto.NewSHA512()
+
+	n := int(testing.AllocsPerRun(10, func() {
+		md5Hash.Write(msg)
+		sha1Hash.Write(msg)
+		sha256Hash.Write(msg)
+		sha512Hash.Write(msg)
+
+		md5Hash.Sum(nil)
+		sha1Hash.Sum(nil)
+		sha256Hash.Sum(nil)
+		sha512Hash.Sum(nil)
+
+		md5Hash.Reset()
+		sha1Hash.Reset()
+		sha256Hash.Reset()
+		sha512Hash.Reset()
+	}))
+	want := 8
+	if compareCurrentVersion("go1.24") >= 0 {
+		// The go1.24 compiler is able to optimize the allocation away.
+		// See cgo_go124.go for more information.
+		want = 4
```

Comment:
We are we still allocating? Could it be that we need to pass a preallocated buffer (as big as the hash size) to `Sum`?

Reply:
It is same in stdlib's hash too.